File: cosmoauthorization/src/java/it/csi/cosmo/cosmoauthorization/business/service/EntiService.java
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

package it.csi.cosmo.cosmoauthorization.business.service;

import it.csi.cosmo.cosmoauthorization.dto.rest.AggiornaEnteRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.CreaEnteRequest;
import it.csi.cosmo.cosmoauthorization.dto.rest.Ente;
import it.csi.cosmo.cosmoauthorization.dto.rest.EntiResponse;

/**
 *
 */

public interface EntiService {

  /**
   * Metodo per eliminare un ente dal database
   *
   * @param id e' l'identificativo numerico dell'ente da eliminare
   */
  void deleteEnte(Long id);

  /**
   * Metodo che restituisce una lista di enti validi. Questa lista ha un numero di elementi pari al
   * valore indicato da filter o, se filter non e' valorizzato, un numero di elementi pari al valore
   * definito nella tabella di configurazione
   *
   * @param filter indica il numero di enti che deve essere restituito
   * @return una lista di enti validi
   */
  EntiResponse getEnti(String filter);


  /**
   * Metodo che restituisce un ente specifico
   *
   * @param id dell'ente che deve essere cercato nel database
   * @return ente presente nel database
   */
  Ente getEnte(Long id);

  /**
   * Metodo che crea un ente sul database
   *
   * @param ente e' l'ente che deve essere creato sul database
   * @return l'ente creato sul database
   */
  Ente createEnte(CreaEnteRequest ente);

  /**
   * Metodo che aggiorna un ente gia' esistente nel database
   *
   * @param ente e' l'ente che deve essere aggiornato
   * @return l'ente aggiornato nel database
   */
  Ente updateEnte(Long id, AggiornaEnteRequest ente);

}

File: cosmowcl/src/app/shared/services/enti.service.ts
/*
 * Copyright CSI-Piemonte - 2023
 * SPDX-License-Identifier: GPL-3.0-or-later
 */
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { AggiornaEnteRequest } from '../models/api/cosmoauthorization/aggiornaEnteRequest';
import { CreaEnteRequest } from '../models/api/cosmoauthorization/creaEnteRequest';
import { EnteResponse } from '../models/api/cosmoauthorization/enteResponse';
import { EntiResponse } from '../models/api/cosmoauthorization/entiResponse';
import { ApiUrls } from '../utilities/apiurls';
import { Utils } from '../utilities/utilities';

@Injectable()
export class EntiService {

  constructor(
    private http: HttpClient
  ) {
    // NOP
  }

  /** GET: Restituisce un ente. */
  getEnte(id: number): Observable<EnteResponse> {
    Utils.require(id);

    return this.http.get<EnteResponse>(ApiUrls.GET_DELETE_ENTE.replace('{id}', id.toString()));
  }

  /** GET: Restituisce una lista di enti. */
  getEnti(filter: string): Observable<EntiResponse> {
    const options = filter ?
      { params: new HttpParams().set('filter', filter) } : {};
    return this.http.get<EntiResponse>(ApiUrls.ENTI, options);
  }

  /** DELETE: Elimina un ente. */
  deleteEnte(id: number): Observable<EnteResponse> {
    Utils.require(id);

    return this.http.delete<EnteResponse>(ApiUrls.GET_DELETE_ENTE.replace('{id}', id.toString()));
  }

  /**
   * POST: Salva un Ente.
   *
   * @param ente Ente da salvare.
   * @returns Observable<Ente> il ente salvato.
   */
  creaEnte(request: CreaEnteRequest): Observable<EnteResponse> {
    Utils.require(request);

    return this.http.post<EnteResponse>(ApiUrls.ENTI, request);
  }

  /**
   * PUT: Aggiorna un ente.
   *
   * @param ente Ente da aggiornare.
   * @returns Observable<Ente> l'ente aggiornato.
   */

  aggiornaEnte(id: number, request: AggiornaEnteRequest): Observable<EnteResponse> {
    Utils.require(id);
    Utils.require(request);

    return this.http.put<EnteResponse>(ApiUrls.GET_DELETE_ENTE.replace('{id}', id.toString()), request);
  }
}
